util/paused/paused.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package paused implements paused helper functions.
+package paused
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/annotations"
+	v1beta2conditions "sigs.k8s.io/cluster-api/util/conditions/v1beta2"
+	"sigs.k8s.io/cluster-api/util/patch"
+)
+
+type pausedConditionSetter interface {
+	v1beta2conditions.Setter
+	client.Object
+}
+
+// EnsurePausedOption is some configuration that modifies options for a EnsurePausedCondition request.
+type EnsurePausedOption interface {
+	// ApplyToSet applies this configuration to the given EnsurePausedCondition options.
+	ApplyToSet(option *EnsurePausedOptions)
+}
+
+// EnsurePausedOptions allows to define options for the EnsurePausedCondition operation.
+type EnsurePausedOptions struct {
+	targetConditionType string
+}
+
+// ApplyOptions applies the given list options on these options,
+// and then returns itself (for convenient chaining).
+func (o *EnsurePausedOptions) ApplyOptions(opts []EnsurePausedOption) *EnsurePausedOptions {
+	for _, opt := range opts {
+		opt.ApplyToSet(o)
+	}
+	return o
+}
+
+// EnsurePausedCondition sets the paused condition on the object and returns if it should be considered as paused.
+func EnsurePausedCondition[T pausedConditionSetter](ctx context.Context, c client.Client, cluster *clusterv1.Cluster, o T, opts ...EnsurePausedOption) (isPaused bool, conditionChanged bool, err error) {
+	op := &EnsurePausedOptions{
+		targetConditionType: clusterv1.PausedV1Beta2Condition,
+	}
+	for _, opt := range opts {
+		opt.ApplyToSet(op)
+	}
+
+	oldCondition := v1beta2conditions.Get(o, op.targetConditionType)
+	newCondition := pausedCondition(cluster, o, op.targetConditionType)
+
+	isPaused = newCondition.Status == metav1.ConditionTrue
+
+	// Return early if the paused condition did not change.
+	if oldCondition != nil &&
```

Comment:
if we move to the condition package we might consider to use hasSameState

Reply:
hasSameState checks a few more things, we should be really sure about that we want to compare all of them

Reply:
As noted above: we can't move this to the conditions package because we need the patch helper and would have a cyclic dependency.

Exported `HasSameState` instead.